### PR TITLE
Bug fix when state variable is an array

### LINKED ISSUE #446: Issue with array state variables
Consider below mod file:

```
$ cat cabpump.mod
TITLE Calcium ion accumulation and diffusion with pump
: The internal coordinate system is set up in PROCEDURE coord_cadifus()
: and must be executed before computing the concentrations.
: The scale factors set up in this procedure do not have to be recomputed
: when diam or DFree are changed.
: The amount of calcium in an annulus is ca[i]*diam^2*vol[i] with
: ca[0] being the second order correct concentration at the exact edge
: and ca[NANN-1] being the concentration at the exact center

? interface
NEURON {
	SUFFIX cadifpmp
	USEION ca READ cao, ica WRITE cai, ica
	RANGE ica_pmp, ica_pmp_last
	RANGE pump0
}

DEFINE NANN  10

UNITS {
	(mV)	= (millivolt)
	(molar) = (1/liter)
	(mM)	= (millimolar)
	(um)	= (micron)
	(mA)	= (milliamp)
	(mol)	= (1)
	FARADAY = (faraday)	 (coulomb)
	PI	= (pi)		(1)
	R 	= (k-mole)	(joule/degC)
}

PARAMETER {
	DFree = .6	(um2/ms) <0,1e9>
	beta = 50		<0, 1e9>
        k1 = 5e8        (/mM-s) <0, 1e10>:optional mm formulation
        k2 = .25e6      (/s)	<0, 1e10>
        k3 = .5e3       (/s)	<0, 1e10>
        k4 = 5e0        (/mM-s)	<0, 1e10>
	pump0 = 3e-14 (mol/cm2) <0, 1e9> : set to 0 in hoc if this pump not wanted
}

ASSIGNED {
	celsius		(degC)
	diam		(um)
	v		(millivolt)
	cao		(mM)
	cai		(mM)
	ica		(mA/cm2)
        ica_pmp (mA/cm2)
        area1    (um2)
        c1      (1+8 um5/ms)
        c2      (1-10 um2/ms)
        c3      (1-10 um2/ms)
        c4      (1+8 um5/ms)
	ica_pmp_last (mA/cm2)
}

STATE {
	ca[NANN]	(mM) <1e-6> : ca[0] is equivalent to cai
        pump    (mol/cm2)	<1e-15>
        pumpca  (mol/cm2)	<1e-15>
}

BREAKPOINT {
	SOLVE state
	ica_pmp_last = ica_pmp
	ica = ica_pmp
}

PROCEDURE state() {

}
```

MOD2C generates following variables in codegen:

```
#define t _nt->_t
#define dt _nt->_dt
#define pump0 _p[0*_STRIDE]
#define ica_pmp _p[1*_STRIDE]
#define ica_pmp_last _p[2*_STRIDE]
#define ca (_p + 3*_STRIDE)
#define pump _p[13*_STRIDE]
#define pumpca _p[14*_STRIDE]
#define cao _p[15*_STRIDE]
#define cai _p[16*_STRIDE]
#define ica _p[17*_STRIDE]
#define area1 _p[18*_STRIDE]
#define c1 _p[19*_STRIDE]
#define c2 _p[20*_STRIDE]
#define c3 _p[21*_STRIDE]
#define c4 _p[22*_STRIDE]
#define Dca (_p + 23*_STRIDE)
#define Dpump _p[33*_STRIDE]
#define Dpumpca _p[34*_STRIDE]
#define _v_unused _p[35*_STRIDE]
#define _g_unused _p[36*_STRIDE]
```

and NMODL does:

```
    static inline void setup_instance(NrnThread* nt, Memb_list* ml)  {
        cadifpmp_Instance* inst = (cadifpmp_Instance*) mem_alloc(1, sizeof(cadifpmp_Instance));
        int pnodecount = ml->_nodecount_padded;
        Datum* indexes = ml->pdata;
        inst->pump0 = ml->data+0*pnodecount;
        inst->ica_pmp = ml->data+1*pnodecount;
        inst->ica_pmp_last = ml->data+2*pnodecount;
        inst->ca = ml->data+3*pnodecount;
        inst->pump = ml->data+13*pnodecount;
        inst->pumpca = ml->data+14*pnodecount;
        inst->cao = ml->data+15*pnodecount;
        inst->cai = ml->data+16*pnodecount;
        inst->ica = ml->data+17*pnodecount;
        inst->area1 = ml->data+18*pnodecount;
        inst->c1 = ml->data+19*pnodecount;
        inst->c2 = ml->data+20*pnodecount;
        inst->c3 = ml->data+21*pnodecount;
        inst->c4 = ml->data+22*pnodecount;
        inst->Dca = ml->data+23*pnodecount;
        inst->Dpump = ml->data+24*pnodecount;
        inst->Dpumpca = ml->data+25*pnodecount;
        inst->v_unused = ml->data+26*pnodecount;
        inst->g_unused = ml->data+27*pnodecount;
        inst->ion_cao = nt->_data;
```
So we are missing offset for `Dca` variable.

### PR DESCRIPTION
**Pull request by pramodk**
- for each state variable, Dstate variable is allocated
   in the array
 - when calculating vector size, we have to allocate Dstate
   variable as vector and not scalar variable

Todo:
- [x] fix float variables
- [x] fix int variables
- [x] fix ordering of int variables (e.g. _ion_dicadv)

fixes #446

### CODE REVIEW

**Comment by pramodk** [comment]
Ordering of all variables is now consistent between neuron/mod2c and nmodl:

image

File: src/codegen/codegen_c_visitor.cpp [hunk 1]
```diff
@@ -762,7 +762,10 @@ void CodegenCVisitor::update_index_semantics() {
             info.semantics.emplace_back(index++, ion.name + "_ion", 1);
         }
         for (auto& var: ion.writes) {
```
Comment (alkino) [inline src/codegen/codegen_c_visitor.cpp:764]: for (const auto& var: ion.writes) {

File: src/codegen/codegen_c_visitor.cpp [hunk 2]
```diff
@@ -825,6 +828,9 @@ std::vector<SymbolType> CodegenCVisitor::get_float_variables() {
     for (auto& variable: states) {
```
Comment (alkino) [inline src/codegen/codegen_c_visitor.cpp:828]: for (**const** auto& variable: states) {

What about `variable` -> `state`

File: src/codegen/codegen_c_visitor.cpp (hunk 1 above)
Comment (pramodk) [reply src/codegen/codegen_c_visitor.cpp:764]: done

Comment (alkino) [reply src/codegen/codegen_c_visitor.cpp:764]: You fix the wrond one :p

Comment (pramodk) [reply src/codegen/codegen_c_visitor.cpp]: now!

**Review by alkino — APPROVED** [review]

**Pull request merged**